apps/oracles/hyper-borrow-rates/src/providers/euler.rs

```diff
@@ -0,0 +1,160 @@
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+use alloy::primitives::{Address, U256};
+use alloy::providers::ProviderBuilder;
+use alloy::sol_types::SolCall;
+use anyhow::Result;
+use futures::{stream::FuturesUnordered, StreamExt};
+use url::Url;
+
+use crate::domain::{BorrowRateInfo, FeedConfig, RatesPerFeed};
+use crate::providers::onchain::{MyProvider, ETHEREUM_MAINNET_RPC_URL};
+use crate::utils::math::apr_continuous;
+use blocksense_sdk::eth_rpc::eth_call;
+
+pub mod euler {
+    alloy::sol! {
+        #[allow(missing_docs)]
+        #[allow(clippy::too_many_arguments)]
+        #[sol(rpc)]
+        EulerUtilsLens,
+        "src/abi/Euler/utilsLens.json"
+    }
+}
+
+async fn fetch_borrow_rate(
+    provider: &MyProvider,
+    lens_address: Address,
+    vault_address: Address,
+) -> Result<U256> {
+    // let rpc_url = Url::parse(RPC_URL)?;
+    // let provider = ProviderBuilder::new().connect_http(rpc_url.clone());
+
+    let instance = euler::EulerUtilsLens::new(lens_address, provider);
+    let calldata = instance.getAPYs(vault_address).calldata().clone();
+
+    let raw = eth_call(
+        ETHEREUM_MAINNET_RPC_URL,
+        &format!("{:?}", lens_address),
+        &calldata,
+    )
```

Comment:
Why `.clone()`-ing if going to create a reference just after? Does it mess with the lifetimes somehow?

```suggestion
    let calldata = instance.getAPYs(vault_address).calldata();

    let raw = eth_call(
        ETHEREUM_MAINNET_RPC_URL,
        &format!("{:?}", lens_address),
        calldata,
    )
```

Reply:
yep

Reply:
nvm this worked. tnx
```rs
    let call = instance.getAPYs(vault_address);
    let calldata = call.calldata();
```